drivers/media/i2c/ov5647.c

```diff
@@ -442,8 +442,30 @@ static int ov5647_enum_mbus_code(struct v4l2_subdev *sd,
 	return 0;
 }
 
+static int ov5647_set_get_fmt(struct v4l2_subdev *sd,
+			      struct v4l2_subdev_pad_config *cfg,
+			      struct v4l2_subdev_format *format)
+{
+	struct v4l2_mbus_framefmt *fmt = &format->format;
+
+	if (format->pad != 0)
+		return -EINVAL;
+
+	/* Only one format is supported, so return that */
+	memset(fmt, 0, sizeof(*fmt));
+	fmt->code = MEDIA_BUS_FMT_SBGGR8_1X8;
+	fmt->colorspace = V4L2_COLORSPACE_SRGB;
+	fmt->field = V4L2_FIELD_NONE;
+	fmt->width = 640;
+	fmt->height = 480;
+
+	return 0;
+}
+
 static const struct v4l2_subdev_pad_ops ov5647_subdev_pad_ops = {
 	.enum_mbus_code = ov5647_enum_mbus_code,
+	.set_fmt =	  ov5647_set_get_fmt,
+	.get_fmt =	  ov5647_set_get_fmt,
```

Comment:
set_fmt is obliged to do the best it can to match the requested format/resolution.
In this case there is only one format/resolution, therefore the answer is always the same, and is the same as getting the format.

Reply:
So it isn't strange that the parameters supplied to the `set` get overwritten?

Reply:
No, not in this case.
Should the driver support multiple modes then it'd get updated to choose the best match, but often in sensor driver such as this where only certain modes are supported, that will still result in an overwrite of the input.

Docs at https://www.kernel.org/doc/html/v4.15/media/uapi/v4l/vidioc-g-fmt.html and https://www.kernel.org/doc/html/v4.15/media/uapi/v4l/vidioc-subdev-g-fmt.html#vidioc-subdev-g-fmt (Main V4L2 and subdev APIs respectively.)
V4L2 docs section 7.31.4 para 3:
> When the application calls the VIDIOC_S_FMT ioctl with a pointer to a struct v4l2_format structure the driver checks and adjusts the parameters against hardware abilities. Drivers should not return an error code unless the type field is invalid, this is a mechanism to fathom device capabilities and to approach parameters acceptable for both the application and driver. On success the driver may program the hardware, allocate resources and generally prepare for data exchange. Finally the VIDIOC_S_FMT ioctl returns the current format parameters as VIDIOC_G_FMT does. **Very simple, inflexible devices may even ignore all input and always return the default parameter**